File: Aula_Ricardo/Programação_orientada_a_objetos/python/Sobrecarga_de_metodos.py
class Produto:
    def __init__(self, nome, valor):
        self.nome = nome
        self.valor = valor 

    def __add__(self, other):
        if isinstance(other, Produto):
            return Produto(self.nome+ ' & ' + other.nome, self.valor + other.valor)
        return NotImplemented
    
    def __str__(self):
        return f'{self.nome}: R${self.valor:.2f}'
    

produto1 = Produto("Produto A", 22)
produto2 = Produto("Produto B", 33)
produto3 = Produto("Produto C", 65)

produto4 = produto1 + produto2 + produto3

print(produto4)

File: Aula_Ricardo/Programação_orientada_a_objetos/go/Sobrecarga_de_metodos.go
package main

import "fmt"

type Produto struct {
	Nome  string
	Valor float64
}

func (p Produto) Adicionar(other Produto) Produto {
	return Produto{
		Nome:  p.Nome + " & " + other.Nome,
		Valor: p.Valor + other.Valor,
	}
}

func (p Produto) String() string {
	return fmt.Sprintf("%s: R$%.2f", p.Nome, p.Valor)
}

func main() {
	produto1 := Produto{Nome: "Produto A", Valor: 22.0}
	produto2 := Produto{Nome: "Produto B", Valor: 33.0}
	produto3 := Produto{Nome: "Produto C", Valor: 65.0}

	produto4 := produto1.Adicionar(produto2).Adicionar(produto3)

	fmt.Println(produto4)
}
